### PR TITLE
PyTree: Pass node_data to tree walkers

### PR DESCRIPTION
**Pull request by awf**
Add an optional argument to `PyTreeDef::Walk` which causes `node_data` to be passed to the node callback, allowing that callback to inspect the node type.

In particular, it allows the definition of custom printers (see https://github.com/google/jax/issues/5047), which can e.g. print a tree with path labels like:
```
/a1: leaf-a1
/a2/b1/0: 111
/a2/b1/1: 2
/a2/b1/2: 3
/a2/b2: [[1. 1. 1.]\n [1. 1. 1.]]
/a2/b3/c1: 33
/a2/b3/c2/0: 44
/a2/b3/c2/1: 55
/a3/x1/x11: a
/a3/x1/x12/y1/y11: ya
/a3/x1/x12/y1/y12: yb
/a3/x1/x13: b
/a3/x2/x21: 2a
/a3/x2/x22/y1/y11: ya
/a3/x2/x22/y1/y12: yb
/a4: fred
```
corresponding to tree `obj` in the new tests (see below).

Having just seen the comment that `Walk` is to be deprecated, perhaps an alternative is to make `PyTreeDef`s a little more introspectable, e.g. add something like `treedef.child_names()`.
The downside of the latter idea is that it's not clear how to define it for Custom nodes, 
whereas the node_data is exactly the auxiliary data supplied by the programmer.

I would be happy to implement any proposal that enables printing as above.

@hawkinsp you might be willing to offer a view?

### CODE REVIEW

**Comment by gbaned** [comment]
Hi @r4nt Can you please review this PR ? Thank you!

File: tensorflow/compiler/xla/python/BUILD [hunk 1]
```diff
@@ -82,6 +82,18 @@ py_test(
     ] + xla_py_test_deps(),
 )
 
+py_test(
```
Comment (hawkinsp) [inline tensorflow/compiler/xla/python/BUILD:85]: I will note that the existing tests for this functionality are in the JAX tree:
https://github.com/google/jax/blob/main/tests/tree_util_test.py

I might be tempted to add your new tests there instead. You have to do it as a separate PR; the way you make a change across both repositories is that you bump the client version number in xla_client.py in this repository (in this directory), and then gate your new test on that version (exposed in jax as `jax._src.lib.xla_extension_version`).

File: tensorflow/compiler/xla/python/pytree.cc [hunk 2]
```diff
@@ -491,7 +491,8 @@ py::list PyTreeDef::FlattenUpTo(py::handle xs) const {
 }
 
 py::object PyTreeDef::Walk(const py::function& f_node, py::handle f_leaf,
-                           py::iterable leaves) const {
+                           py::iterable leaves,
+                           bool pass_node_data) const {
```
Comment (hawkinsp) [inline tensorflow/compiler/xla/python/pytree.cc:495]: Given this isn't an API we advertise or document, I might be tempted just to change the API contract of f_node to unconditionally pass the node data.

The only use of this function I can find anywhere is this one: https://github.com/google/jax/blob/be41c8c1d3ab79078acd9113f35e179a4ebf8f33/jax/_src/tree_util.py#L195
and as best I can tell, that function is completely unused. So I might just delete it.

**Review by hawkinsp — COMMENTED** [review]
I'm sorry I didn't even see this PR until now! There's a much higher risk I will miss PRs sent to the TF tree since I don't check for those in my email.

File: tensorflow/compiler/xla/python/pytree.cc (hunk 2 above)
Comment (awf) [reply tensorflow/compiler/xla/python/pytree.cc:495]: Thanks! That was my first design, so I'll happily revert to it.  I wasn't sure if there might be keen `Walk`ers in other parts of the ecosystem.

**Comment by awf** [comment]
> I'm sorry I didn't even see this PR until now! There's a much higher risk I will miss PRs sent to the TF tree since I don't check for those in my email.

No worries, I had a workaround.

> I might be tempted to add your new tests there instead. 

Will do.

File: tensorflow/compiler/xla/python/pytree.h [hunk 3]
```diff
@@ -144,8 +144,7 @@ class PyTreeDef {
   std::vector<std::unique_ptr<PyTreeDef>> Children() const;
 
   // Maps a function over a PyTree structure, applying f_leaf to each leaf, and
-  // f_node to each container node.
-  // TODO(phawkins): use flattening everywhere instead and delete this method.
```
Comment (awf) [inline tensorflow/compiler/xla/python/pytree.h:148]: I'm wondering about removing this, as this now provides functionality that flattening doesn't?

**Review by hawkinsp — APPROVED** [review]

**Comment by awf** [comment]
Do I need to do anything on this?  I'm guessing not, but just checking...
